File: crwno/BOJ/boj2512.py
# N = int(input())
# budget = list(map(int, input().split()))
# M = int(input())
#
# # sum = 0
# ans = 0
# k = M // N
# for i in budget:
#     sum += i
# if sum <= M:
#     print(max(budget))
# else:
#     while M - ans > 0:
#         k += 1
#         ans = 0
#         for i in budget:
#             if i < k:
#                 ans += i
#             else:
#                 ans += k
#     else:
#         print(k - 1)

# sum = 0
# ans = 0
# k = M // N
# h = 0
# for i in budget:
#     sum += i
# if sum <= M:
#     print(max(budget))
# else:
#     while (M - ans) // N > 0:
#         k += h
#         ans = 0
#         for i in budget:
#             if i < k:
#                 ans += i
#             else:
#                 ans += k
#         h = (M - ans) // N
#
# while M - ans > 0:
#     k += 1
#     ans = 0
#     for i in budget:
#         if i < k:
#             ans += i
#         else:
#             ans += k
# else:
#     print(k - 1)

N = int(input())
budget = list(map(int, input().split()))
M = int(input())

sum1 = sum(budget)
if sum1 <= M:
    print(max(budget))
else:
    low, high = 0, M
    result = 0

    while low <= high:
        mid = (low + high) // 2
        total = 0

        for i in budget:
            total += min(i, mid)

        if total <= M:
            result = mid
            low = mid + 1
        else:
            high = mid - 1

    print(result)

File: aeongseo/BOJ/CPP/boj2512.cpp
/*** 2512. 예산 ***/

// 예산 총액 넘으면 상한액 감소
// 예산 총액 넘지 않으면 상한액 증가

#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;

vector<int> cost;

int main() {
	int N, budget, ans;
	cin >> N;
	
	cost.assign(N, 0);

	int maxcost = 0;
	for (int i = 0; i < N; i++) {
		cin >> cost[i];
		maxcost = max(maxcost, cost[i]);
	}

	cin >> budget;

	int left = 0, right = maxcost;
	while (left <= right) {
		int mid = (left + right) / 2;
		int sum = 0;
		
		for (int i = 0; i < N; i++) {
			sum += min(cost[i], mid);
		}

		if (sum == budget) {
			ans = mid;
			break;
		}
		else if (sum < budget) {
			ans = mid; 
			left = mid + 1;
		}
		else right = mid - 1;
	}
	
	cout << ans;

	return 0;
}